stl/inc/xloctime

```diff
@@ -651,6 +651,17 @@ protected:
     __CLR_OR_THIS_CALL ~time_get_byname() noexcept override {}
 };
 
+constexpr bool _Is_valid_fmt_specifier(const char _Specifier) {
+    constexpr char _Valid_specifiers[] = "aAbBcCdDeFgGhHIjmMnprRStTuUVwWxXyYzZ";
+    for (const char _Valid_specifier : _Valid_specifiers) {
+        if (_Valid_specifier == _Specifier) {
+            return true;
+        }
+    }
+
+    return false;
+}

```

Comment:
I was thinking about чstd::binary_searchч here, but I guess this would require movement to a .cpp file which as of now does not exist for чxloctimeч. Also not sure this is the correct place for this function in general. I'm accepting any guidelines or suggestions.

P.S. Also should I have __CLRCALL_OR_CDECL here?

Reply:
> P.S. Also should I have __CLRCALL_OR_CDECL here?

IIUC, no. This macro is DLL-related, and we shoudn't touch the DLLs just for this issue.

It can be still fine to move `_Is_valid_fmt_specifier` to an import library. But as the function body is very simple, I think it's better to make it header only.

Reply:
Things should be injected into the import lib only when necessary (e.g. WinAPI calls) or there is a *very* compelling throughput reason (e.g. charconv tables). Import lib additions are "eternal" and they are subject to strict constraints (can't drag in arbitrary stuff like non-core headers), so we don't want to add them arbitrarily.

The basic rule is that header-only code is happy code. Separately compiled code is weird and scary. I learned these lessons at great cost.

Reply:
If you want the runtime perf, `std::binary_earch` isn't even the best strategy here.

The best is to have 256-char byte-bool or bit mask pre-computed. Though I'm not sure if it fits here, as `static constexpr` is questioned below. But with `_Is_constant_evaluated()` should be doable for runtime only.

Another good strategy may be to implement it like `if (_Specifier == 'a' || _Specifier == 'A' || ...)`. the compiler is likely to convert char ranges to machine word bitsets, chechk them with `bt` instruction, and do fewer steps that `std::binary_earch` i

Reply:
Then I guess instead of `if (_Specifier == 'a' || _Specifier == 'A' || ...)` it would be even better to do switch/case table?

Reply:
>  it would be even better to do switch/case table?

Yes. 
Unfortunately MSVC can'te emit for switch/case anything better than jump table. Other compilers do,
See https://godbolt.org/z/sbo6oEqnj

Although both are ugly in the source, so it depends on if the optimization is justified here.

Computing a `static const` bitmap or bytemap is also good (IMO, better that `if (...||...)`, and not ugly in the source.

Reply:
I don't know. @StephanTLavavej are optimization justified here?

Reply:
Just to mention that largest bottleneck of the `std::put_time` is the call to strftime which for the non-wchar version will  convert the string to wchar internally, then fill the result and convert back to char and this is done for each and every specifier in a loop.

In case we don't want to rewrite the implementation to not use `strftime` at all, a much better runtime optimization would be to prepare the whole string in advance with properly escaping characters that would cause `strftime` to crash and then call it only once. If we do that, the optimization for the loop here will be completely not worth it.

Reply:
Ok, agree

Reply:
Yeah, I find the "`strftime` is the bottleneck" rationale persuasive, so I'm not too concerned about optimizing this code, even though a table would be an efficient strategy when considered in isolation. This approach is fine unless and until we look into optimizing the code as a whole (i.e. let's worry about correctness and performance separately).